packages/hub/src/hooks/useMutationForm.ts

```diff
@@ -0,0 +1,69 @@
+import { FieldValues, UseFormProps, useForm } from "react-hook-form";
+import { VariablesOf } from "relay-runtime";
+
+import { CommonMutationParameters, useAsyncMutation } from "./useAsyncMutation";
+
+/**
+ * This hook ties together `useForm` and `useAsyncMutation`, which is a very common pattern for forms that submit data to our backend through mutations.
+ * It forwards some of its parameters to `useForm`, and others to `useAsyncMutation`, and converts form's data to mutation's variables with `formDataToVariables` function.
+ *
+ * See also:
+ * - `<MutationModalAction>` if your form is available through a Dropdown menu
+ * - `useAsyncMutation` if your mutation doesn't require a form, and for the details on parameters for this function
+ *
+ * All generic type parameters to this function default to `never`, so you'll have to set them explicitly to pass type checks.
+ */
+export function useMutationForm<
+  FormShape extends FieldValues = never,
+  TMutation extends CommonMutationParameters<TTypename> = never,
+  TTypename extends string = never,
+>({
+  // useForm params
+  defaultValues,
+  mode,
+  // useAsyncMutation params
+  mutation,
+  expectedTypename,
+  blockOnSuccess,
+  confirmation,
+  // runMutation params
+  onCompleted,
+  // bridge from form to runMutation
+  formDataToVariables,
+}: {
+  // This is unfortunately not strictly type-safe: if you return extra variables that are not needed for mutation, TypeScript won't complain.
+  // See also: https://stackoverflow.com/questions/72111571/typescript-exact-return-type-of-function
+  // This could be solved by converting the return type to generic, but I expect that the lack of partial type parameters in TypeScript
+  // would get in the way, so I won't even try.
+  formDataToVariables: (data: FormShape) => VariablesOf<TMutation>;
+} & Pick<UseFormProps<FormShape>, "defaultValues" | "mode"> &
+  Pick<
+    Parameters<typeof useAsyncMutation<TMutation, TTypename>>[0],
+    // I could just pass everything but I'm worried about potential name collisions, and it's easy to whitelist everything necessary.
+    // We have to list all these when we pass them to `useAsyncMutation`, anyway.
+    "mutation" | "expectedTypename" | "blockOnSuccess" | "confirmation"
+  > &
+  // TODO - with `useAsyncMutation`, submitted data is available because `onCompleted` is usually a closure over it.
+  // For `useMutationForm`, it'd be useful to pass it to `onCompleted` explicitly, `onCompleted(result, submittedData)`.
+  Pick<
+    Parameters<ReturnType<typeof useAsyncMutation<TMutation, TTypename>>[0]>[0],
+    "onCompleted"
```

Comment:
This is the most complex type in this PR, so I'll explain it as an example of what's going on and why.

`useAsyncMutation` returns `runMutation` function, which you can call to actually run the mutation. With this hook, it runs on form submit.

`runMutation` takes `variables` parameter (almost always), but also `onCompleted` callback to run if the mutation succeeded.

In this hook, though, since `runMutation` is invoked when the form is submitted, `onCompleted` callback needs to be passed in immediately when "mutation form" is created.

`onCompleted` type is complicated: its parameter is `OkResult`, defined inside `useAsyncMutation`. It's inferred from the mutation result type (generated by Relay), but filtered for `expectedTypename`.

I could write out the full `onCompleted` function type here, and it would work, but that would be a copy-paste from `useAsyncFunction`. So instead, I infer the type from `useAsyncMutation` type itself.

- `typeof useAsyncMutation<TMutation, TTypename>` — this limits `useAsyncMutation` to its specialized non-generic version
- `ReturnType<...>[0]` extracts `runMutation` type
- `Parameters<...>[0]` extracts `runMutation`'s first parameter
- `Pick<..., "onCompleted">` extract `{ onCompleted }` type